[Win] Support USE_TYPO_METRICS flag

fujii opened the pull request:
#### 0df29c6b290b4cc961c34097fb02b7d2c8c491e1
<pre>
[Win] Support USE_TYPO_METRICS flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=275892">https://bugs.webkit.org/show_bug.cgi?id=275892</a>

Reviewed by Frédéric Wang.

Windows WebKit was using incorrect font metrics for Google Sans font
on Google News. The font uses USE_TYPO_METRICS flag.

If USE_TYPO_METRICS flag is set in fsSelection of OS/2 table, The font
should use typo metrics in the table.
&lt;<a href="https://learn.microsoft.com/en-us/typography/opentype/spec/os2">https://learn.microsoft.com/en-us/typography/opentype/spec/os2</a>&gt;
&gt; If set, it is strongly recommended that applications use
&gt; OS/2.sTypoAscender - OS/2.sTypoDescender + OS/2.sTypoLineGap as the
&gt; default line spacing for this font.

USE_TYPO_METRICS flag support was added by
&lt;<a href="https://commits.webkit.org/168985@main">https://commits.webkit.org/168985@main</a>&gt; for Windows port once.
However, it was reverted by &lt;<a href="https://commits.webkit.org/213183@main">https://commits.webkit.org/213183@main</a>&gt;
for three reasons. 1) The descent was wrongly calculated. The negative
value should have been inverted. 2) The test case
fonts/use-typo-metrics-1.html didn&apos;t pass. Windows API
GetOutlineTextMetrics doesn&apos;t preserve USE_TYPO_METRICS flag for
OpenType fonts. 3) I didn&apos;t know actual web pages having layout
problems.

Google Sans font on Google News is using the USE_TYPO_METRICS flag,
and GetOutlineTextMetrics preserves the flag of the font. Restored the
original implementation with fixing the negative descent problem.
Added a new test case using a TrueType font.

* LayoutTests/fonts/resources/use-typo-metrics-2.py: Added.
* LayoutTests/fonts/resources/use-typo-metrics-2.ttf: Added.
* LayoutTests/fonts/use-typo-metrics-2-expected.html: Added.
* LayoutTests/fonts/use-typo-metrics-2.html: Added.
* Source/WebCore/platform/graphics/win/SimpleFontDataWin.cpp:

Canonical link: <a href="https://commits.webkit.org/280480@main">https://commits.webkit.org/280480@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8831ec4f9fb8da155238d45335ab11b824610d0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60360 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/7188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58872 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43696 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7378 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/7188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58775 "Failed to checkout and rebase branch from PR 30257") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/48979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/26825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/30684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/6310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/6191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/6581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/62043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/658 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/62043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/661 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/49034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/62043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8443 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/31902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/32987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/34072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32733 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang reviewed (approved):
s/nagative/negative/
s/preverse/preserve/
s/preseves/preserves/
s/nagative/negative/

LGTM.

It's weird that GetOutlineTextMetrics doesn't preverse USE_TYPO_METRICS flag for
OpenType fonts. Maybe we should also use typo metrics if a MATH table is present so that math fonts work on Windows, but anyway that's a separate issue...

Maybe you should also leave a comment on bug 199186 after this patch lands, to explain the situation.

fujii committed:
[Win] Support USE_TYPO_METRICS flag
https://bugs.webkit.org/show_bug.cgi?id=275892

Reviewed by Frédéric Wang.

Windows WebKit was using incorrect font metrics for Google Sans font
on Google News. The font uses USE_TYPO_METRICS flag.

If USE_TYPO_METRICS flag is set in fsSelection of OS/2 table, The font
should use typo metrics in the table.
<https://learn.microsoft.com/en-us/typography/opentype/spec/os2>
> If set, it is strongly recommended that applications use
> OS/2.sTypoAscender - OS/2.sTypoDescender + OS/2.sTypoLineGap as the
> default line spacing for this font.

USE_TYPO_METRICS flag support was added by
<https://commits.webkit.org/168985@main> for Windows port once.
However, it was reverted by <https://commits.webkit.org/213183@main>
for three reasons. 1) The descent was wrongly calculated. The negative
value should have been inverted. 2) The test case
fonts/use-typo-metrics-1.html didn't pass. Windows API
GetOutlineTextMetrics doesn't preserve USE_TYPO_METRICS flag for
OpenType fonts. 3) I didn't know actual web pages having layout
problems.

Google Sans font on Google News is using the USE_TYPO_METRICS flag,
and GetOutlineTextMetrics preserves the flag of the font. Restored the
original implementation with fixing the negative descent problem.
Added a new test case using a TrueType font.

* LayoutTests/fonts/resources/use-typo-metrics-2.py: Added.
* LayoutTests/fonts/resources/use-typo-metrics-2.ttf: Added.
* LayoutTests/fonts/use-typo-metrics-2-expected.html: Added.
* LayoutTests/fonts/use-typo-metrics-2.html: Added.
* Source/WebCore/platform/graphics/win/SimpleFontDataWin.cpp:

Canonical link: https://commits.webkit.org/280480@main
